File: lib/domain/cultivation_plan/gateways/plan_allocation_allocate_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      class PlanAllocationAllocateGateway
        # @raise [Domain::CultivationPlan::Errors::AllocationNoCandidatesError]
        # @raise [Domain::CultivationPlan::Errors::AllocationExecutionError]
        def allocate(fields:, crops:, weather_data:, planning_start:, planning_end:, interaction_rules: nil, objective: "maximize_profit", max_time: nil, enable_parallel: false)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_allocation_allocate_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::PlanAllocationAllocateGateway`

use serde_json::Value;
use time::Date;

pub trait PlanAllocationAllocateGateway: Send + Sync {
    #[allow(clippy::too_many_arguments)]
    fn allocate(
        &self,
        fields: &[Value],
        crops: &[Value],
        weather_data: &Value,
        planning_start: Date,
        planning_end: Date,
        interaction_rules: Option<&Value>,
        objective: &str,
        max_time: Option<i64>,
        enable_parallel: bool,
    ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>>;
}
